environs/instances/instancetype.go

```diff
@@ -210,3 +211,29 @@ func (s byMemory) Less(i, j int) bool {
 	// Result is in descending order of cost so instance with lowest cost is used.
 	return inst0.Cost > inst1.Cost
 }
+
+//by Type is used to sort a slice by name by best effort. As we have different separators for different providers
+//we do it by best effort using "-" and "."
+//A possible sort could be:
+//a1.2xlarge, a1.large, lexical sort would put a1.2xlarge before a1.large -- aws
+
+type ByName []InstanceType
+
+func (bt ByName) Len() int      { return len(bt) }
+func (bt ByName) Swap(i, j int) { bt[i], bt[j] = bt[j], bt[i] }
+func (bt ByName) Less(i, j int) bool {
+	inst0, inst1 := &bt[i], &bt[j]
+	baseInst0 := strings.FieldsFunc(inst0.Name, splitDelimiters)
+	baseInst1 := strings.FieldsFunc(inst1.Name, splitDelimiters)
+	if baseInst0[0] != baseInst1[0] {
```

Comment:
Can you be sure that `baseInst0` and `baseInst0` are greater than 0, if the `.Name` of either is empty, then this will panic for slice bound issue. You should add a check there.

Reply:
go strings package: `If s does not contain sep and sep is not empty, Split returns a slice of length 1 whose only element is s.`
Tested if locally, even if nothing is found the return is always a len(slice) == 1.